File: src/2022-intern/712.minimum-ascii-delete-sum-for-two-strings.js
/*
 * @lc app=leetcode id=712 lang=javascript
 *
 * [712] Minimum ASCII Delete Sum for Two Strings
 */

// @lc code=start
/**
 * @param {string} word1
 * @param {string} word2
 * @return {number}
 */
var minimumDeleteSum = function (word1, word2) {
  /**
   * dp[i][j] word1[0,i-1] 和 word2[0,j-1] 所需的最小删除 ASCII 和
   *
   * base case
   * dp[i][0] = sum(word1.charCodeAt(0...i-1))
   * dp[0][j] = sum(word2.charCodeAt(0...j-1))
   *
   * if(word1[i-1] === word2[j-1])
   *  dp[i][j] = dp[i-1][j-1]
   * else
   *  dp[i][j] = Math.min(
   *      dp[i][j-1] + word2.charCodeAt(j-1) // 留下 word1[i-1]，删掉 word2[j-1]
   *      dp[i-1][j] +word1.charCodeAt(i-1)// 留下 word2[j-1]，删掉 word1[i-1]
   * )
   *
   * return dp[word1.length][word2.length]
   */

  const dp = new Array(word1.length + 1);
  for (let i = 0; i <= word1.length; i++) {
    dp[i] = new Array(word2.length + 1);
  }

  dp[0][0] = 0;
  let word1ASCIISum = 0;
  for (let i = 1; i <= word1.length; i++) {
    word1ASCIISum += word1.charCodeAt(i - 1);
    dp[i][0] = word1ASCIISum;
  }

  let word2ASCIISum = 0;
  for (let j = 1; j <= word2.length; j++) {
    word2ASCIISum += word2.charCodeAt(j - 1);
    dp[0][j] = word2ASCIISum;
  }

  for (let i = 1; i <= word1.length; i++) {
    for (let j = 1; j <= word2.length; j++) {
      if (word1[i - 1] === word2[j - 1]) {
        dp[i][j] = dp[i - 1][j - 1];
      } else {
        dp[i][j] = Math.min(
          // 留下 word1[i-1]
          dp[i][j - 1] + word2.charCodeAt(j - 1),
          // 留下 word2[j-1]
          dp[i - 1][j] + word1.charCodeAt(i - 1),
        );
      }
    }
  }

  return dp[word1.length][word2.length];
};
// @lc code=end

File: src/2026/712.minimum-ascii-delete-sum-for-two-strings.cpp
/*
 * @lc app=leetcode id=712 lang=cpp
 *
 * [712] Minimum ASCII Delete Sum for Two Strings
 */

#include <numeric>
#include <string>
#include <vector>

// @lc code=start
class Solution {
 public:
  int minimumDeleteSum(const std::string& s1, const std::string& s2) {
    const int total_ascii_sum = std::accumulate(s1.cbegin(), s1.cend(), 0) +
                                std::accumulate(s2.cbegin(), s2.cend(), 0);

    // dp[i][j] - Common sequence with maximum ASCII sum in s1[:i-1] and
    // s2[:j-1]
    // Base case
    // dp[0][...] = dp[...][0] = 0
    std::vector<std::vector<int>> dp(2, std::vector<int>(s2.size() + 1, 0));

    for (int i = 1; i <= s1.size(); i++) {
      for (int j = 1; j <= s2.size(); j++) {
        dp[i % 2][j] = std::max(
            {// Ignore s1[i-1]
             dp[(i - 1) % 2][j],
             // Ignore s2[j-1]
             dp[i % 2][j - 1],
             // Ignore both if s1[i-1] != s2[i-1], or include both if s1[i-1] ==
             // s2[j-1]
             dp[(i - 1) % 2][j - 1] + (s1[i - 1] == s2[j - 1]) * s1[i - 1]});
      }
    }

    // Use total ASCII to remove the common sequence with maximum ASCII sum to
    // get minimum deleted ASCII sum.
    const int result = total_ascii_sum - 2 * dp[(s1.size()) % 2][s2.size()];
    return result;
  }
};
// @lc code=end

int main() {
  Solution sol;
  sol.minimumDeleteSum("sea", "eat");
}